app/src/main/java/org/mozilla/focus/web/GeckoWebViewProvider.kt

```diff
@@ -0,0 +1,598 @@
+        }
+
+        override fun setBlockingEnabled(enabled: Boolean) {
+            geckoSession.settings.setBoolean(GeckoSessionSettings.USE_TRACKING_PROTECTION, enabled)
+            if (enabled) {
+                updateBlocking()
+                applyAppSettings()
+            } else {
+                geckoRuntime!!.settings.javaScriptEnabled = true
+                geckoRuntime!!.settings.webFontsEnabled = true
+                geckoRuntime!!.settings.cookieBehavior = GeckoRuntimeSettings.COOKIE_ACCEPT_ALL
+            }
+            callback?.onBlockingStateChanged(enabled)
+        }
+
+        override fun setRequestDesktop(shouldRequestDesktop: Boolean) {
+            geckoSession.settings.setBoolean(
+                GeckoSessionSettings.USE_DESKTOP_MODE,
+                shouldRequestDesktop
+            )
+            callback?.onRequestDesktopStateChanged(shouldRequestDesktop)
+        }
+
+        override fun onSharedPreferenceChanged(
+            sharedPreferences: SharedPreferences,
+            prefName: String
+        ) {
+            updateBlocking()
+            applyAppSettings()
+        }
+
+        private fun applyAppSettings() {
+            geckoRuntime!!.settings.javaScriptEnabled =
+                    !Settings.getInstance(context).shouldBlockJavaScript()
+            geckoRuntime!!.settings.webFontsEnabled =
+                    !Settings.getInstance(context).shouldBlockWebFonts()
+            geckoRuntime!!.settings.remoteDebuggingEnabled = false
+            val cookiesValue = if (Settings.getInstance(context).shouldBlockCookies() &&
+                Settings.getInstance(context).shouldBlockThirdPartyCookies()
+            ) {
+                GeckoRuntimeSettings.COOKIE_ACCEPT_NONE
+            } else if (Settings.getInstance(context).shouldBlockThirdPartyCookies()) {
+                GeckoRuntimeSettings.COOKIE_ACCEPT_FIRST_PARTY
+            } else {
+                GeckoRuntimeSettings.COOKIE_ACCEPT_ALL
+            }
+            geckoRuntime!!.settings.cookieBehavior = cookiesValue
+        }
+
+        private fun updateBlocking() {
+            val settings = Settings.getInstance(context)
+
+            var categories = 0
+            if (settings.shouldBlockSocialTrackers()) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_SOCIAL
+            }
+            if (settings.shouldBlockAdTrackers()) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_AD
+            }
+            if (settings.shouldBlockAnalyticTrackers()) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_ANALYTIC
+            }
+            if (settings.shouldBlockOtherTrackers()) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_CONTENT
+            }
+
+            geckoRuntime!!.settings.trackingProtectionCategories = categories
+        }
+
+        @Suppress("ComplexMethod")
```

Comment:
\[detekt\] INFO: \[style/ReturnCount\] Restrict the number of return statements in methods\.